m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,781 @@
+    private String durationBucketString(Duration bucketBound) {
+        if (Duration.ZERO.equals(bucketBound)) {
+            return "0";
+        }
+
+        if (Duration.MAX_VALUE.equals(bucketBound)) {
+            return "infinity";
+        }
+
+        if (Duration.MIN_VALUE.equals(bucketBound)) {
+            return "-infinity";
+        }
+
+        return bucketBound.toString();
+    }
+
+    @Override
+    public void reportCounter(String name, Map<String, String> tags, long value) {
+        synchronized (isOpenLock) {
+            if (!isOpen) {
+                LOG.error("Reporter already closed; no more metrics can be flushed");
+                return;
+            }
+
+            reportCounterInternal(name, tags, value);
+        }
+    }
+
+    @Override
+    public void reportGauge(String name, Map<String, String> tags, double value) {
+        synchronized (isOpenLock) {
+            if (!isOpen) {
+                LOG.error("Reporter already closed; no more metrics can be flushed");
+                return;
+            }
+
+            GaugeValue gaugeValue = new GaugeValue();
+            gaugeValue.setDValue(value);
+
+            MetricValue metricValue = new MetricValue();
+            metricValue.setGauge(gaugeValue);
+
+            Metric metric = newMetric(name, tags, metricValue);
+
+            queueSizedMetric(new SizedMetric(metric, calculateSize(metric)));
+        }
+    }
+
+    @Override
+    public void reportTimer(String name, Map<String, String> tags, Duration interval) {
+        synchronized (isOpenLock) {
+            if (!isOpen) {
+                LOG.error("Reporter already closed; no more metrics can be flushed");
+                return;
+            }
+
+            TimerValue timerValue = new TimerValue();
+            timerValue.setI64Value(interval.getNanos());
+
+            MetricValue metricValue = new MetricValue();
+            metricValue.setTimer(timerValue);
+
+            Metric metric = newMetric(name, tags, metricValue);
+
+            queueSizedMetric(new SizedMetric(metric, calculateSize(metric)));
+        }
+    }
+
+    @Override
+    public void reportHistogramValueSamples(
+        String name,
+        Map<String, String> tags,
+        Buckets buckets,
+        double bucketLowerBound,
+        double bucketUpperBound,
+        long samples
+    ) {
+        synchronized (isOpenLock) {
+            if (!isOpen) {
+                LOG.error("Reporter already closed; no more metrics can be flushed");
+                return;
+            }
+
+            // Append histogram bucket-specific tags
+            int bucketIdLen = String.valueOf(buckets.size()).length();
+            bucketIdLen = Math.max(bucketIdLen, MIN_METRIC_BUCKET_ID_TAG_LENGTH);
+
+            String bucketIdFmt = String.format("%%0%sd", bucketIdLen);
+
+            BucketPair[] bucketPairs = BucketPairImpl.bucketPairs(buckets);
+
+            if (tags == null) {
+                tags = new HashMap<>(2, 1);
```

Comment:
what does the `(2,1)` here mean?

Reply:
Removed this optimization for readability.